src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp

```diff
@@ -3071,9 +3068,167 @@ void OptoRuntime::generate_exception_blob() {
+   : StubCodeGenerator(buffer, PrintMethodHandleStubs),
+     _call_target(call_target),
+     _shadow_space_bytes(shadow_space_bytes),
+     _input_registers(input_registers),
+     _output_registers(output_registers) {}
+  void generate();
+
+private:
+#ifdef ASSERT
+  bool target_uses_register(VMReg reg) {
+    return _input_registers.contains(reg) || _output_registers.contains(reg);
+  }
+#endif
+};
+
+static const int native_invoker_code_size = 1024;
+
 BufferBlob* SharedRuntime::make_native_invoker(address call_target,
-                                           int shadow_space_bytes,
-                                           const GrowableArray<VMReg>& input_registers,
-                                           const GrowableArray<VMReg>& output_registers) {
-  return NULL;
+                                               int shadow_space_bytes,
+                                               const GrowableArray<VMReg>& input_registers,
+                                               const GrowableArray<VMReg>& output_registers) {
+  BufferBlob* _invoke_native_blob =
+    BufferBlob::create("nep_invoker_blob", native_invoker_code_size);
+  if (_invoke_native_blob == NULL)
+    return NULL; // allocation failure
+
+  CodeBuffer code(_invoke_native_blob);
+  NativeInvokerGenerator g(&code, call_target, shadow_space_bytes, input_registers, output_registers);
+  g.generate();
+  code.log_section_sizes("nep_invoker_blob");
+
+  return _invoke_native_blob;
+}
+
+void NativeInvokerGenerator::generate() {
+  assert(!(target_uses_register(rscratch1->as_VMReg())
+           || target_uses_register(rscratch2->as_VMReg())
+           || target_uses_register(rthread->as_VMReg())),
+         "Register conflict");
+
+  MacroAssembler* masm = _masm;
+
+  __ set_last_Java_frame(sp, noreg, lr, rscratch1);
+
+  __ enter();
+
+  // Store a pointer to the previous R29 (RFP) saved on the stack as it
+  // may contain an oop if PreserveFramePointer is off. This value is
+  // retrieved later by frame::sender_for_entry_frame() when the stack
+  // is walked.
+  __ mov(rscratch1, sp);
+  __ str(rscratch1, Address(rthread, JavaThread::saved_fp_address_offset()));
+
+  // State transition
+  __ mov(rscratch1, _thread_in_native);
+  __ strw(rscratch1, Address(rthread, JavaThread::thread_state_offset()));
+
+  assert(_shadow_space_bytes == 0, "not expecting shadow space on AArch64");
+
+  rt_call(masm, _call_target);
+
+  __ mov(rscratch1, _thread_in_native_trans);
+  __ strw(rscratch1, Address(rthread, JavaThread::thread_state_offset()));
+
+  // Force this write out before the read below
+  __ membar(Assembler::LoadLoad | Assembler::LoadStore |
+            Assembler::StoreLoad | Assembler::StoreStore);
+
+  __ verify_sve_vector_length();
+
+  Label L_after_safepoint_poll;
+  Label L_safepoint_poll_slow_path;
+
+  __ safepoint_poll(L_safepoint_poll_slow_path, rthread, true /* at_return */, false /* in_nmethod */);
+
```

Comment:
This looks wrong: please look at the definition of MacroAssembler::safepoint_poll, which has an acquire flag.

Reply:
Yes this is bad: it generates the correct code but the arguments are totally wrong.